javascript/ql/src/semmle/javascript/security/dataflow/UnsafeJQueryPluginCustomizations.qll

```diff
@@ -0,0 +1,242 @@
+  }
+
+  /**
+   * Gets an operand to `extend`.
+   */
+  private DataFlow::SourceNode getAnExtendOperand(DataFlow::TypeBackTracker t, ExtendCall extend) {
+    t.start() and
+    result.flowsTo(extend.getAnOperand())
+    or
+    exists(DataFlow::TypeBackTracker t2 | result = getAnExtendOperand(t2, extend).backtrack(t2, t))
+  }
+
+  /**
+   * A function that is registered as a jQuery plugin method.
+   */
+  class JQueryPluginMethod extends DataFlow::FunctionNode {
+    string pluginName;
+
+    JQueryPluginMethod() {
+      exists(DataFlow::Node def |
+        jQueryPluginDefinition(pluginName, def) and
+        this = getAJQueryPluginMethod(DataFlow::TypeBackTracker::end(), def)
+      )
+    }
+
+    /**
+     * Gets the name of this plugin.
+     */
+    string getPluginName() { result = pluginName }
+  }
+
+  /**
+   * Holds if `plugin` has a default option defined a `def`.
+   */
+  private predicate hasDefaultOption(JQueryPluginMethod plugin, DataFlow::PropWrite def) {
+    exists(ExtendCall extend, JQueryPluginOptions options, DataFlow::SourceNode default |
+      options.getPlugin() = plugin and
+      options = getAnExtendOperand(DataFlow::TypeBackTracker::end(), extend) and
+      default = getAnExtendOperand(DataFlow::TypeBackTracker::end(), extend) and
+      default.getAPropertyWrite() = def
+    )
+  }
+
+  /**
+   * The client-provided options object for a jQuery plugin.
+   */
+  class JQueryPluginOptions extends DataFlow::ParameterNode {
+    JQueryPluginMethod method;
+
+    JQueryPluginOptions() {
+      exists(string optionsPattern |
+        optionsPattern = "(?i)(opt(ion)?s?)" and
+        if method.getAParameter().getName().regexpMatch(optionsPattern)
+        then (
+          // use the last parameter named something like "options" if it exists ...
+          getName().regexpMatch(optionsPattern) and
+          this = method.getAParameter()
+        ) else (
+          // ... otherwise, use the last parameter, unless it looks like a DOM node
+          this = method.getLastParameter() and
+          not getName().regexpMatch("(?i)(e(l(em(ent(s)?)?)?)?)")
+        )
+      )
+    }
+
+    /**
+     * Gets the plugin method that these options are used in.
+     */
+    JQueryPluginMethod getPlugin() { result = method }
+  }
+
+  /**
+   * Expression of form `isElement(x)`, which sanitizes `x`.
+   */
+  class IsElementSanitizer extends TaintTracking::SanitizerGuardNode, DataFlow::CallNode {
+    IsElementSanitizer() {
+      // common ad hoc sanitizing calls
+      exists(string name | getCalleeName() = name |
+        name = "isElement" or name = "isWindow" or name = "isWindow"
+      )
+    }
+
+    override predicate sanitizes(boolean outcome, Expr e) {
+      outcome = true and e = getArgument(0).asExpr()
+    }
+  }
+
+  /**
+   * Expression of form `typeof x.jquery !== "undefined"` or `x.jquery`, which sanitizes `x`.
```

Comment:
It also matches expressions of the form `x.jquery !== undefined`, maybe that case should also be mentioned in the docstring.